enterprise/coderd/prebuilds/membership_test.go

```diff
@@ -32,87 +38,168 @@ func TestReconcileAll(t *testing.T) {
+									})
+								}
+							}
+
+							presets := []database.GetTemplatePresetsWithPrebuildsRow{newPresetRow(unrelatedOrg.ID)}
+							if includePreset {
+								presets = append(presets, newPresetRow(targetOrg.ID))
+							}
+
+							// Verify memberships before reconciliation.
+							preReconcileMemberships, err := db.GetOrganizationsByUserID(ctx, database.GetOrganizationsByUserIDParams{
+								UserID: database.PrebuildsSystemUserID,
+							})
+							require.NoError(t, err)
+							expectedMembershipsBefore := []uuid.UUID{defaultOrg.ID, unrelatedOrg.ID}
+							if preExistingOrgMembership {
+								expectedMembershipsBefore = append(expectedMembershipsBefore, targetOrg.ID)
+							}
+							require.ElementsMatch(t, expectedMembershipsBefore, extractOrgIDs(preReconcileMemberships))
+
+							// Reconcile
+							reconciler := prebuilds.NewStoreMembershipReconciler(db, clock)
+							require.NoError(t, reconciler.ReconcileAll(ctx, database.PrebuildsSystemUserID, presets))
+
+							// Verify memberships after reconciliation.
+							postReconcileMemberships, err := db.GetOrganizationsByUserID(ctx, database.GetOrganizationsByUserIDParams{
+								UserID: database.PrebuildsSystemUserID,
+							})
+							require.NoError(t, err)
+							expectedMembershipsAfter := expectedMembershipsBefore
+							if !preExistingOrgMembership && tc.expectOrgMembershipExists != nil && *tc.expectOrgMembershipExists {
+								expectedMembershipsAfter = append(expectedMembershipsAfter, targetOrg.ID)
+							}
+							require.ElementsMatch(t, expectedMembershipsAfter, extractOrgIDs(postReconcileMemberships))
+
+							// Verify prebuilds group behavior based on expected outcomes
+							prebuildsGroup, err = db.GetGroupByOrgAndName(ctx, database.GetGroupByOrgAndNameParams{
+								OrganizationID: targetOrg.ID,
+								Name:           "prebuilds",
+							})
+							if tc.expectGroupExists != nil && *tc.expectGroupExists {
+								require.NoError(t, err)
+								require.Equal(t, "prebuilds", prebuildsGroup.Name)
+								require.Equal(t, "Prebuilds", prebuildsGroup.DisplayName)
+								require.Equal(t, int32(0), prebuildsGroup.QuotaAllowance) // Default quota should be 0
+
+								if tc.expectUserInGroup != nil && *tc.expectUserInGroup {
+									// Check that the system user is a member of the prebuilds group
+									groupMembers, err := db.GetGroupMembersByGroupID(ctx, database.GetGroupMembersByGroupIDParams{
+										GroupID:       prebuildsGroup.ID,
+										IncludeSystem: true,
+									})
+									require.NoError(t, err)
+									require.Len(t, groupMembers, 1)
+									require.Equal(t, database.PrebuildsSystemUserID, groupMembers[0].UserID)
+								}
+
+								if tc.expectUserInGroup != nil && !*tc.expectUserInGroup {
+									// Check that the system user is NOT a member of the prebuilds group
+									groupMembers, err := db.GetGroupMembersByGroupID(ctx, database.GetGroupMembersByGroupIDParams{
```

Comment:
Shouldn't the prebuilds user always be part of the prebuilds group? Or is this the case where an organization doesn't have prebuilds?

Reply:
It is not invariant in the first test case, which is half of the total test space.
We test the negative case, where there are no presets, to ensure that we do not add the user to orgs or groups that don't need it.